app/src/main/java/com/example/healthinspector/Fragments/ScanFlow/RecommendProductsFragment.java

```diff
@@ -0,0 +1,176 @@
+        ArrayList<RecommendedProduct> recommendedProducts = new ArrayList<>();
+        String url = "";
+        url += URL_REQUEST_PRODUCTS;
+        //limiting the categories to 4, since 5 is too much and most of the time won't give back results
+        int categoriesCount = scannedProduct.getCategories().size();
+        if(categoriesCount == 0){returnToProductDetails(scannedProduct);} //if there are no categories, then no recommendations can be made for this scanned product
+        if(categoriesCount > 4){categoriesCount = 4;}
+        for(int i = 0; i < categoriesCount; i++){
+            String category = "&" + "tagtype_" + i + "=categories&tag_contains_" + i + "=contains&tag_" + i + "=" + scannedProduct.getCategories().get(i);
+            url+= category;
+        }
+
+        String healthyProductParams = "&tagtype_" + (categoriesCount) + "=nutrition_grades&tag_contains_" + (categoriesCount) + "=contains&tag_" + (categoriesCount) + "=A&additives=without&ingredients_from_palm_oil=without&json=true";
+        url += healthyProductParams;
+        //new volley request to get product recommendations
+        RequestQueue queue = Volley.newRequestQueue(requireContext());
+        JsonObjectRequest recommendedProductsRequest = new JsonObjectRequest(Request.Method.GET, url, null,
+                new Response.Listener<JSONObject>() {
+                    @Override
+                    public void onResponse(JSONObject response) {
+                        int count = 0;
+                        try {
+                            count = response.getInt(COUNT);
+                        } catch (JSONException e) {
+                            e.printStackTrace();
+                        }
+                        //if the response has 0 products, return to details
+                        if(count == 0){
+                            returnToProductDetails(scannedProduct);
+                            return;
+                        }
+                        //limiting to 10 different recommendations
+                        if(count > 10){count = 10;}
+                        for(int i = 0; i < count; i++){
+                            String productName = "";
+                            String brand = "";
+                            String keywords = "";
+
+                            try {
+                                productName = response.getJSONArray(PRODUCTS).getJSONObject(i).getString(Constants.PRODUCT_NAME);
+
+                                if(response.getJSONArray(PRODUCTS).getJSONObject(i).has(BRAND)){
+                                    brand = response.getJSONArray(PRODUCTS).getJSONObject(i).getString(BRAND);
+                                }
+
+                                StringBuilder keywordsBuilder =new StringBuilder();
+                                JSONArray productKeywords = response.getJSONArray(PRODUCTS).getJSONObject(i).getJSONArray(KEYWORDS);
```

Comment:
same comment as earlier diff about storing things like `response.getJSONArray(PRODUCTS).getJSONObject(i)` in a variable so you are not repeating code